File: driver/py/ID1000500B.py
import logging, time
from ipdi.ip.pyaip import pyaip, pyaip_init
import sys

## IP Convolution driver class
class convolution:
    ## Class constructor of IP Convolution driver
    #
    # @param self Object pointer
    # @param targetConn Middleware object
    # @param config Dictionary with IP Convolution configs
    # @param addrs Network address IP
    def __init__(self, connector, nic_addr, port, csv_file):
        ## Pyaip object
        self.__pyaip = pyaip_init(connector, nic_addr, port, csv_file)
        if self.__pyaip is None:
            logging.debug("error")
        ## Array of strings with information read
        self.__pyaip.reset()
        self.dataRX = []
        ## IP Convolution IP-ID
        self.IPID = 0
        self.__getID()
        self.__clearStatus()
        logging.debug(f"IP Convolution controller created with IP ID {self.IPID:08x}")

    ## Convolution Process
    #
    # @param self Object pointer
    # @param Y    Array of data to put in memory Y
    def conv(self, memY):          
        # Enable interruptions
        # self.__enableINT()
        # Write Conf Reg with size of memory Y
        self.__writeSizeY(len(memY))
        # Write memory: MEM_Y
        self.__writeData(memY)
        # Start IP
        self.__startIP()
        # Show status
        self.__status()
        # Clear Interrups
        #self.__disableINT()
        # Wait for Done Flag
        self.__waitInt()
        print('end waiting')
        # Read memory: MEM_Z
        memZ = self.__readData(64)
        # Clear flags
        self.__clearStatus()           
        # Show status
        self.__status()
        
        return memZ

    ## Set size of memory Y in IP Convolution processing
    #
    # @param self Object pointer
    # @param sizeY size of memory Y
    def __writeSizeY(self, sizeY):

        parameter = []
        parameter.append(sizeY)
        self.__pyaip.writeConfReg('CONFREG', parameter, 1, 0)
        logging.debug(f"Confreg setted to {sizeY}")
        
    ## Write data in the IP Convolution input memory
    #
    # @param self Object pointer
    # @param data String array with the data to write
    def __writeData(self, data):
        self.__pyaip.writeMem('MEM_Y', data, len(data), 0)
        logging.debug("Data captured in MEM_Y")
        print('Write memory: MEM_Y')
        print(f'memY Data: {[f"{x:08X}" for x in memY]}\n')        

    ## Start processing in IP Convolution
    #
    # @param self Object pointer
    def __startIP(self):
        print('startIP')
        self.__pyaip.start()
        logging.debug("Start sent")
    
    ## Wait for the completion of the process
    #
    # @param self Object pointer
    def __waitInt(self):
        print('Waiting')
        waiting = True        
        while waiting:
            status = self.__pyaip.getStatus()
            ##print('Status: ', status)
            logging.debug(f"status {status:08x}")            
            if status & 0x1:
                waiting = False            
            time.sleep(0.1)
    
    ## Read data from the IP Convolution output memory
    #
    # @param self Object pointer
    # @param size Amount of data to read
    def __readData(self, size):
        data = self.__pyaip.readMem('MEM_Z', size, 0)
        logging.debug("Data obtained from MemZ")
        print(f'result Data: {[f"{x:08X}" for x in data]}\n')
        return data   
    
    ## Show IP Convolution STATUS
    #
    # @param self Object pointer
    def __status(self):
        print('here status')
        status = self.__pyaip.getStatus()
        logging.info(f"{status:08x}")   
    
    ## Clear status register of IP Convolution
    #
    # @param self Object pointer
    def __clearStatus(self):
        for i in range(8):
            self.__pyaip.clearINT(i)

    ## Enable IP Convolution interruptions
    #
    # @param self Object pointer
    def __enableINT(self):
        self.__pyaip.enableINT(0,None)
        logging.debug("Int enabled")

    ## Disable IP Convolution interruptions
    #
    # @param self Object pointer
    def __disableINT(self):
        self.__pyaip.disableINT(0)

        logging.debug("Int disabled")  

    ## Get IP ID
    #
    # @param self Object pointer
    def __getID(self):
        self.IPID = self.__pyaip.getID()

    ## Finish connection
    #
    # @param self Object pointer
    def finish(self):
        self.__pyaip.finish()           
        
if __name__=="__main__":
    import sys, random, time, os
    logging.basicConfig(level=logging.DEBUG)

    aip_mem_size = 8
    size_mem_y   = 5
    connector = '/dev/ttyACM0'
    nic_addr = 1
    port = 0
    csv_file = '/home/pablo/convo_core/ID1000500B_config.csv'
    convolutioner = 0

    try:       
        convolutioner = convolution(connector, nic_addr, port, csv_file)
        ##convolutioner.reset()
    except:
        e = sys.exc_info()
        print('ERROR: ', e)
    
    #==========================================
    #random.seed(64)
    #memY = [random.randrange(2 ** 8) for i in range(0, size_mem_y)]
    memY = [ 0, 1, 2, 3, 4]
    memZ = convolutioner.conv(memY)
    convolutioner.finish()
    logging.info("DONE!")

    # ID = convolutioner.getID()
    # print(f'Read ID: {ID:08X}\n')
    
    
    # sizeY = [0x00000005]
    # print('Write configuration register: CONFREG')
    # convolutioner.writeConfReg('CONFREG', sizeY, 1, 0)
    # print(f'sizeY Data: {[f"{x:08X}" for x in sizeY]}\n')
    
    # memY = [random.randrange(2 ** 32) for i in range(0, aip_mem_size)]
    # convolutioner.writeData(memY)
   
    

    # memY = [0x00000000, 0x00000001, 0x00000002, 0x00000003, 0x00000004]

    # print('Write memory: MEM_Y')
    # convolutioner.writeMem('MEM_Y', memY, 5, 0)
    # print(f'memY Data: {[f"{x:08X}" for x in memY]}\n')

    # print('Start IP\n')
    # convolutioner.start()

    # STATUS = convolutioner.getStatus()
    # print(f'Read STATUS: {STATUS:08X}\n')

    # print('Read memory: MEM_Z')
    # result = convolutioner.readMem('MEM_Z', 63, 0)
    # print(f'result Data: {[f"{x:08X}" for x in result]}\n')

    # print('Clear INT: 0')
    # convolutioner.clearINT(0)

    # STATUS = convolutioner.getStatus()
    # print(f'Read STATUS: {STATUS:08X}\n')

    #==========================================

    # convolutioner.finish()


    # convolutioner.finish()
        

File: driver/c/id1000500b.c
#include <stdio.h>
#include <stdbool.h>
#include "id1000500b.h"
#include "caip.h"

#ifdef _WIN32
#include <windows.h>
#else
#include <unistd.h>
#endif // _WIN32

/** DEFINITIONS*/
#define SUCCESS     0
#define FAULT       -1

#define INT_DONE    0
#define ONE_FLIT    1
#define ZERO_OFFSET 0
#define STATUS_BITS 8
#define INT_DONE_BIT    0x00000001

/** Global variables declaration (private) */
caip_t      *id1000500b_aip;
uint32_t    id1000500b_id = 0;
/***********************************************************************/
static uint32_t id1000500b_getID(uint32_t  *id);
static uint32_t id1000500b_clearStatus(void);



/***********************************************************************/

/* Initializes the aip to run */
int32_t id1000500b_init(const char *connector, uint8_t nic_addr, uint8_t port, const char *csv_file){
    
    id1000500b_aip = caip_init(connector, nic_addr, port, csv_file);

    if(id1000500b_aip == NULL){
        printf("CAIP Object not created");
        return FAULT;
    }
    id1000500b_aip->reset();

    id1000500b_getID(&id1000500b_id);
    id1000500b_clearStatus();

    printf("\nIP Convolution Core controller created with IP ID: %08X\n\n", id1000500b_id);
    return SUCCESS;
}

// WRITE DATA
int32_t id1000500b_writeData(uint32_t *data, uint32_t dataSize){

    //El mnemonico del primer parametro debe coincidir con nuestro csv_file de configuracion
    id1000500b_aip->writeMem("MEM_Y", data, dataSize, ZERO_OFFSET);
    return SUCCESS;
}

/* Write the length of the input array */
int32_t id1000500b_writeConfReg(uint32_t sizeX){
    
    id1000500b_aip->writeConfReg("CONFREG", &sizeX, ONE_FLIT, ZERO_OFFSET);
    return SUCCESS;
}

/* Read data*/
int32_t id1000500b_readData(uint32_t *data, uint32_t data_size)
{
    
    id1000500b_aip->readMem("MEM_Z", data, data_size, ZERO_OFFSET);
    return SUCCESS;
}

/* Start processing*/
int32_t id1000500b_startIP(void){

    id1000500b_aip->start();
    return SUCCESS;

}

/* Enable interruption notification "Done"*/
int32_t id1000500b_enableINT(void)
{
    id1000500b_aip->enableINT(INT_DONE, NULL);
    printf("\nINT Done enabled");
    return SUCCESS;
}

/* Disable interruption notification "Done"*/
int32_t id1000500b_disableINT(void)
{
    id1000500b_aip->disableINT(INT_DONE);
    printf("\nINT Done disabled");
    return SUCCESS;
}

/* Show status*/
int32_t id1000500b_status(void)
{
    uint32_t status;
    id1000500b_aip->getStatus(&status);
    printf("\nStatus: %08X",status);
    return SUCCESS;
}

/* Wait interruption*/
int32_t id1000500b_waitINT(void)
{
    bool waiting = true;
    uint32_t status;

    while(waiting)
    {
        id1000500b_aip->getStatus(&status);

        if((status & INT_DONE_BIT)>0)
            waiting = false;

        #ifdef _WIN32
        Sleep(500); // ms
        #else
        sleep(0.1); // segs
        #endif
    }

    id1000500b_aip->clearINT(INT_DONE);

    return 0;
}

/* Finish*/
int32_t id1000500b_finish(void)
{
    id1000500b_aip->finish();
    return SUCCESS;
}

//PRIVATE FUNCTIONS
uint32_t id1000500b_getID(uint32_t *id)
{
    id1000500b_aip->getID(id);

    return 0;
}

uint32_t id1000500b_clearStatus(void)
{
    for(uint8_t i = 0; i < STATUS_BITS; i++)
        id1000500b_aip->clearINT(i);

    return 0;
}